File: 백준/Gold/2493. 탑/탑.py
import sys

input = sys.stdin.readline
stack = []
top = -1
answer = []

def isEmpty():
    if len(stack)== 0:
        return True
    else:
        return False
if __name__ == "__main__":
    N = int(input())

    towers = list(map(int, input().split()))
    top+=1
    stack.append([towers[0], 1])
    answer.append(0)
    for i in range(1, N):
        while(not isEmpty()):
            if stack[top][0] > towers[i]:
                break
            top-=1
            stack.pop()
        if isEmpty():
            answer.append(0)
        else:
            answer.append(stack[top][1])
        stack.append([towers[i], i+1])
        top+=1
    
    for i in range(len(answer)):
        print(answer[i],end=" ")

File: 백준/Gold/2493. 탑/탑.cc
#include <bits/stdc++.h> 
using namespace std;

#define X first
#define Y second
int towers[500000];
int answers[500000];

int main() {
	ios::sync_with_stdio(0);
	cin.tie(0);
	stack<pair<int, int>> tower;

	int n;
	cin >> n;

	for (int i = 0; i < n; i++) {
		cin >> towers[i];
	}

	for (int i = 0; i < n; i++) {
		// tower.push({ towers[i - 1], i});

		while (!tower.empty() && tower.top().X < towers[i]) {
			tower.pop();
		}
		if (tower.empty()) {
			answers[i] = 0;
		}
		else {
			answers[i] = tower.top().Y;
		}
		tower.push({ towers[i], i+1});
	}

	for (int i = 0; i < n; i++)
		cout << answers[i] << " ";
}
